### PR TITLE
packagekit: Install auto update backend on demand

### PR DESCRIPTION
**Pull request by Gundersanne**

### CODE REVIEW

File: pkg/packagekit/autoupdates.jsx [hunk 1]
```diff
@@ -203,16 +209,15 @@ class DnfImpl extends ImplBase {
...
         let dfd = cockpit.defer();
         getBackend.promise = dfd.promise();
 
-        // TODO: only check for dnf/yum/apt and install the "automatic" packages dynamically
-        cockpit.script(["if rpm -q dnf-automatic >/dev/null; then echo dnf; " +
-                        "elif rpm -q yum-cron >/dev/null; then echo yum; " +
-                        "elif dpkg -s unattended-upgrades >/dev/null 2>&1; then echo apt; fi"],
+        cockpit.script(["(command -v dnf >/dev/null && echo dnf) || " +
+                        "(command -v yum >/dev/null && echo yum) || " +
+                        "(command -v apt >/dev/null && echo apt)"],
```
Comment (martinpitt) [inline pkg/packagekit/autoupdates.jsx]: Completely nerdy suggestion: `command -v dnf yum apt  | head -n1 | xargs basename`, or do the head/basename in JS.

File: pkg/packagekit/autoupdates.jsx [hunk 2]
```diff
@@ -275,6 +288,17 @@ export default class AutoUpdates extends React.Component {
         if (!backend)
             return null;
 
+        if (!backend.installed) {
+            return (
+                <div className="header-buttons pk-updates--header pk-updates--header--auto" id="automatic">
+                    <h2 className="pk-updates--header--heading">{_("Automatic Updates")}</h2>
+                    <div className="pk-updates--header--actions">
+                        <button className="btn btn-default"
+                                onClick={() => { install_dialog(backend.packageName).then(() => this.initializeBackend(true), () => null) }}>{_("Enable")}</button>
```
Comment (martinpitt) [inline pkg/packagekit/autoupdates.jsx]: Do we really need a new UI for this? I thought clicking the On/Off switch could bring up the install dialog. It seems a bit confusing to click on "Enable" and then auto-updates are still "Off".

File: test/verify/check-packagekit [hunk 3]
```diff
@@ -1035,19 +1035,62 @@ class TestAutoUpdates(PackageCase):
...
+            '/usr/lib/systemd/system/dnf-automatic-install.timer': timerContent
+        });
+        self.enableRepo()
 
+        m.start_cockpit()
+        b.login_and_go('/updates')
+
+        # click through install dialog
+        b.wait_present('#automatic')
+        b.click('#automatic button.btn')
```
Comment (martinpitt) [inline test/verify/check-packagekit]: This needs to wait for the button to be present. (Also, it should be the On/Off switch, see above).

File: test/verify/check-packagekit [hunk 4]
```diff
@@ -1035,19 +1035,62 @@ class TestAutoUpdates(PackageCase):
...
+        b.wait_popup('dialog')
+        b.wait_visible('#dialog button.apply')
+        b.wait_not_attr('#dialog button.apply', 'disabled', '')
```
Comment (martinpitt) [inline test/verify/check-packagekit]: Should be redundant. `click()` will refuse to operate on disabled elements.

File: pkg/packagekit/autoupdates.jsx [hunk 5]
```diff
@@ -203,16 +209,15 @@ class DnfImpl extends ImplBase {
 
 // Returns a promise for instantiating "backend"; this will never fail, if
 // automatic updates are not supported, backend will be null.
-function getBackend() {
-    if (!getBackend.promise) {
-        debug("getBackend() called first time, initializing promise");
+function getBackend(clearPromise) {
```
Comment (martinpitt) [inline pkg/packagekit/autoupdates.jsx]: `clearPromise` - took me a bit to understand what this does. Perhaps `forceReinit` or `noCache`?

Comment (Gundersanne) [reply pkg/packagekit/autoupdates.jsx]: yea the name wasn't the best. noCache/forceReinit is a lot clearer. Will change.

File: test/verify/check-packagekit (hunk 4 above)
Comment (Gundersanne) [reply test/verify/check-packagekit]: If you call click on a disabled element it'll fail though

File: pkg/packagekit/autoupdates.jsx (hunk 2 above)
Comment (Gundersanne) [reply pkg/packagekit/autoupdates.jsx]: That makes a lot more sense, will do it that way.

File: pkg/packagekit/autoupdates.jsx (hunk 1 above)
Comment (Gundersanne) [reply pkg/packagekit/autoupdates.jsx]: Haha ok, that's a lot more onelinery

File: test/verify/check-packagekit [hunk 6]
```diff
@@ -1035,19 +1035,62 @@ class TestAutoUpdates(PackageCase):
...
 
         # empty state visible in main area
         b.wait_present(".container-fluid div.blank-slate-pf")
         self.assertFalse(b.is_present("#available"))
-        # automatic updates are not visible
-        self.assertFalse(b.is_present("#automatic"))
+        # switch to enable automatic upgrades is not visible
+        self.assertFalse(b.is_present("#automatic div.btn-onoff-ct"))
+
+    @skipImage("No supported auto update backend", "centos-7", "debian-stable", "debian-testing", "ubuntu-1604", "ubuntu-stable", "rhel-7-5")
```
[the remark this answers is not in the document]
Comment (Gundersanne) [reply test/verify/check-packagekit]: I'll add those

File: test/verify/check-packagekit [hunk 7]
```diff
@@ -1035,19 +1035,62 @@ class TestAutoUpdates(PackageCase):
...
+    def testInstall(self):
+        b = self.browser
+        m = self.machine
+
+        m.execute('dnf remove -y dnf-automatic')
+
+        # provide minimal content in order for the backend to be seen as supported
+        timerContent = '''
+        [Unit]
```
[the remark this answers is not in the document]
Comment (Gundersanne) [reply test/verify/check-packagekit]: Ah will outdent it then

File: test/verify/check-packagekit (hunk 4 above)
Comment (martinpitt) [reply test/verify/check-packagekit]: Ignore me, it's a wait, not an assertion.

File: test/verify/check-packagekit [hunk 8]
```diff
@@ -1035,19 +1036,62 @@ class TestAutoUpdates(PackageCase):
...
 
         # empty state visible in main area
         b.wait_present(".container-fluid div.blank-slate-pf")
         self.assertFalse(b.is_present("#available"))
-        # automatic updates are not visible
-        self.assertFalse(b.is_present("#automatic"))
+        # switch to enable automatic upgrades is not visible
+        self.assertFalse(b.is_present("#automatic div.btn-onoff-ct"))
+
+    @skipImage("No supported auto update backend", "centos-7", "debian-stable", "debian-testing", "ubuntu-1604", "ubuntu-stable", "rhel-7-5", "rhel-7-5-distropkg", "rhel-7-6")
```
Comment (martinpitt) [inline test/verify/check-packagekit]: ... and now also `rhel-7-6-distropkg`, sorry :)

File: test/verify/check-packagekit [hunk 9]
```diff
@@ -1021,11 +1021,12 @@ class TestAutoUpdates(PackageCase):
         self.assertNotIn("vanilla", out)
         self.assertNotIn("Shutdown", out)
 
+    @skipImage("No supported auto update backend", "centos-7", "debian-stable", "debian-testing", "ubuntu-1604", "ubuntu-stable", "rhel-7-5", "rhel-7-5-distropkg", "rhel-7-6")
```
Comment (martinpitt) [inline test/verify/check-packagekit]: The point of this test is to see what happens on OSes that don't have a supported auto update backend. So it should not be skipped there.

File: test/verify/check-packagekit [hunk 10]
```diff
@@ -1021,11 +1021,12 @@ class TestAutoUpdates(PackageCase):
...
     def testUnsupported(self):
         b = self.browser
         m = self.machine
 
-        m.execute("if type dnf; then dnf remove -y dnf-automatic; elif type apt; then dpkg -P unattended-upgrades; fi")
```
Comment (martinpitt) [inline test/verify/check-packagekit]: Why does this need to change? I know we don't currently implement an auto-update backend for apt, so it should also not be offered to get enabled; `#automatic` should continue to not be visible there, i. e. what the test currently does.

File: pkg/packagekit/autoupdates.jsx [hunk 11]
```diff
@@ -60,20 +63,23 @@ class ImplBase {
...
         // - the config file determines whether to apply security updates only
         // - by default this runs every day (OnUnitInactiveSec=1d), but the timer can be changed with a timer unit
         //   drop-in, so get the last line
-        cockpit.script("if systemctl --quiet is-enabled dnf-automatic-install.timer 2>/dev/null || " +
+        cockpit.script("if rpm -q " + this.packageName + " >/dev/null; then echo installed; fi; " +
+                       "if systemctl --quiet is-enabled dnf-automatic-install.timer 2>/dev/null || " +
                        "  (systemctl --quiet is-enabled dnf-automatic.timer 2>/dev/null && grep -q '^[ \t]*apply_updates[ \t]*=[ \t]*yes' " +
                        "    /etc/dnf/automatic.conf); then echo enabled; fi; " +
-                       "if grep -q '^[ \\t]*upgrade_type[ \\t]*=[ \\t]*security' /etc/dnf/automatic.conf; then echo security; fi; " +
+                       "if [ -f /etc/dnf/automatic ] && grep -q '^[ \\t]*upgrade_type[ \\t]*=[ \\t]*security' /etc/dnf/automatic.conf; then echo security; fi; " +
```
Comment (martinpitt) [inline pkg/packagekit/autoupdates.jsx]: This is to avoid an error message of grep in the logs?

**Comment by Gundersanne** [comment]
On platforms where `dnf-automatic` is available: 
image

**Pull request merged**